### PR TITLE
UseVfsMixin: add vf_trust parameter

### PR DESCRIPTION
**Pull request by jtluka**
### Description

This replaces original messed up PR https://github.com/LNST-project/lnst/pull/379

This adds vf_trust parameter to UseVfsMixin. The Device class is extended with vf_trust() method.
Tests

This also updates BondRecipe, VlansOverBondRecipe and DoubleBondRecipe with fail_over_mac. To avoid code duplication I refactored these classes to us new BondingMixin class that adds the common bonding recipe parameters and provides a method to create the bonding devices.

J:10034309

Reviews

@olichtne

### CODE REVIEW

**Comment by jtluka** [comment]
VF status can be fetched using:
```
[root@wsfd-advnetlab197 ~]# cat /sys/class/net/eno12399/device/sriov_numvfs 
0
[root@wsfd-advnetlab197 ~]# echo 3 > /sys/class/net/eno12399/device/sriov_numvfs 
[root@wsfd-advnetlab197 ~]# cat /sys/class/net/eno12399/device/sriov_numvfs 
3
[root@wsfd-advnetlab197 ~]# ip l show dev eno12399
4: eno12399: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether b4:83:51:01:ea:c4 brd ff:ff:ff:ff:ff:ff
    vf 0     link/ether d2:a3:e7:73:a5:f6 brd ff:ff:ff:ff:ff:ff, spoof checking on, link-state auto, trust off
    vf 1     link/ether 1e:53:d7:e2:e2:fd brd ff:ff:ff:ff:ff:ff, spoof checking on, link-state auto, trust off
    vf 2     link/ether c2:52:5e:b9:0d:a9 brd ff:ff:ff:ff:ff:ff, spoof checking on, link-state auto, trust off
    altname enp49s0f0

```

**Comment by olichtne** [comment]
> VF status can be fetched using:
> 
> ```
> [root@wsfd-advnetlab197 ~]# cat /sys/class/net/eno12399/device/sriov_numvfs 
> 0
> [root@wsfd-advnetlab197 ~]# echo 3 > /sys/class/net/eno12399/device/sriov_numvfs 
> [root@wsfd-advnetlab197 ~]# cat /sys/class/net/eno12399/device/sriov_numvfs 
> 3
> [root@wsfd-advnetlab197 ~]# ip l show dev eno12399
> 4: eno12399: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
>     link/ether b4:83:51:01:ea:c4 brd ff:ff:ff:ff:ff:ff
>     vf 0     link/ether d2:a3:e7:73:a5:f6 brd ff:ff:ff:ff:ff:ff, spoof checking on, link-state auto, trust off
>     vf 1     link/ether 1e:53:d7:e2:e2:fd brd ff:ff:ff:ff:ff:ff, spoof checking on, link-state auto, trust off
>     vf 2     link/ether c2:52:5e:b9:0d:a9 brd ff:ff:ff:ff:ff:ff, spoof checking on, link-state auto, trust off
>     altname enp49s0f0
> ```

can you also try `ip --json l show dev eno12399` to see if this can easily be machine parsed? or if this is somehow accessible from the netlink messages that we already have?

File: lnst/Recipes/ENRT/UseVfsMixin.py [hunk 1]
```diff
@@ -12,19 +13,24 @@ class UseVfsMixin:
...
+    vf_trust = ChoiceParam(choices={'on', 'off'})
 
     def test_wide_configuration(self):
         config = super().test_wide_configuration()
 
         if not self.params.use_vfs:
             return config
 
-        config.vf_config = {}
+        self.vf_config = {}
```
Comment (olichtne) [inline lnst/Recipes/ENRT/UseVfsMixin.py:33]: why the change from `config` to `self`? i would imagine that we may want to export this value into our measurement result metadata to discern between the two possibilities and in that case it should probably be in the shared config object

**Comment by jtluka** [comment]
> ip --json l show dev eno12399

Great idea, it works:
```python
[{"ifindex":4,"ifname":"eno12399","flags":["BROADCAST","MULTICAST"],"mtu":1500,"qdisc":"noop","operstate":"DOWN","linkmode":"DEFAULT","group":"default","txqlen":1000,"link_type":"ether","addr
ess":"b4:83:51:01:ea:c4","broadcast":"ff:ff:ff:ff:ff:ff","vfinfo_list":[{"vf":0,"link_type":"ether","address":"d2:a3:e7:73:a5:f6","broadcast":"ff:ff:ff:ff:ff:ff","vlan_list":[{}],"rate":{"max
_tx":0,"min_tx":0},"spoofchk":true,"link_state":"auto","trust":false},{"vf":1,"link_type":"ether","address":"1e:53:d7:e2:e2:fd","broadcast":"ff:ff:ff:ff:ff:ff","vlan_list":[{}],"rate":{"max_t
x":0,"min_tx":0},"spoofchk":true,"link_state":"auto","trust":false},{"vf":2,"link_type":"ether","address":"c2:52:5e:b9:0d:a9","broadcast":"ff:ff:ff:ff:ff:ff","vlan_list":[{}],"rate":{"max_tx"
:0,"min_tx":0},"spoofchk":true,"link_state":"auto","trust":false}],"altnames":["enp49s0f0"]}]
```

File: lnst/Recipes/ENRT/UseVfsMixin.py [hunk 2]
```diff
@@ -33,17 +39,22 @@ def test_wide_configuration(self):
...
                 sriov_devices = SRIOVDevices(dev, 1)
+
                 vf_dev = sriov_devices.vfs[0]
                 host.map_device(dev_name, {"ifname": vf_dev.name})
 
-                host_config = config.vf_config.setdefault(host, [])
+                host_config = self.vf_config.setdefault(host, [])
                 host_config.append(sriov_devices)
 
+        if self.params.get("vf_trust"):
```
Comment (olichtne) [inline lnst/Recipes/ENRT/UseVfsMixin.py:49]: the `vf_trust` parameter will always have either an `"on"` or `"off"` value so this condition will always be True... maybe you wanted to check `if self.params.use_vfs` but that's also not necessary since you check that higher in the method.

File: lnst/Recipes/ENRT/UseVfsMixin.py [hunk 3]
```diff
@@ -56,10 +67,14 @@ def generate_test_wide_description(self, config):
 
         if self.params.use_vfs:
             description += [
-                f"Using vf device {vf_dev.name} of pf {sriov_devices.phys_dev.name} for DeviceReq {host.hostid}.{vf_dev._id}"
-                for host, sriov_devices_list in config.vf_config.items()
+                f"Using vf device {vf_dev.name} of pf {sriov_devices.phys_dev.name} for DeviceReq {host.hostid}.{vf_dev._id}" + (f" trusted={self.params.vf_trust}" if self.params.get("vf_trust") and sriov_devices.phys_dev in self.vf_trust_device_list else "")
```
Comment (olichtne) [inline lnst/Recipes/ENRT/UseVfsMixin.py:70]: same here... the vf_trust parameter always has a value.

File: lnst/Recipes/ENRT/BondRecipe.py [hunk 4]
```diff
@@ -245,3 +246,7 @@ def pause_frames_dev_list(self):
         """
         return [self.matched.host1.eth0, self.matched.host1.eth1,
             self.matched.host2.eth0]
+
+    @property
+    def vf_trust_device_list(self) -> list[RemoteDevice]:
+        return [sriov_devices.phys_dev for sriov_devices in self.vf_config[self.matched.host1]]
```
Comment (olichtne) [inline lnst/Recipes/ENRT/BondRecipe.py:252]: i think in this case the `sriov_devices` is a `list[Device]` and the `phys_dev` is a `Device` property so you need to nest this one more level:
```
return [device.phys_dev for devices in self.vf_config[self.matched.host1] for device in devices]
```
similar to other recipes...

and probably adjust the variable naming so it's consistent

File: lnst/Recipes/ENRT/DoubleBondRecipe.py [hunk 5]
```diff
@@ -118,3 +119,7 @@ def dev_interrupt_hw_config_dev_list(self):
     def parallel_stream_qdisc_hw_config_dev_list(self):
         host1, host2 = self.matched.host1, self.matched.host2
         return [host1.eth0, host1.eth1, host2.eth0, host2.eth1]
+
+    @property
+    def vf_trust_device_list(self) -> list[RemoteDevice]:
+        return [sriov_devices.phys_dev for sriov_devices_list in self.vf_config.values() for sriov_devices in sriov_devices_list]
```
Comment (olichtne) [inline lnst/Recipes/ENRT/DoubleBondRecipe.py:125]: the nested loop returns a single device i think so probably better to name the variable `sriov_device` instead of `sriov_devices` right?

File: lnst/Recipes/ENRT/VlansOverBondRecipe.py [hunk 6]
```diff
@@ -309,3 +310,7 @@ def pause_frames_dev_list(self):
         """
         host1, host2 = self.matched.host1, self.matched.host2
         return [host1.eth0, host1.eth1, host2.eth0]
+
+    @property
+    def vf_trust_device_list(self) -> list[RemoteDevice]:
+        return [sriov_devices.phys_dev for sriov_devices in self.vf_config[self.matched.host1]]
```
Comment (olichtne) [inline lnst/Recipes/ENRT/VlansOverBondRecipe.py:316]: same problem with list of devices here.

**Review by olichtne — CHANGES_REQUESTED** [review]

File: lnst/Recipes/ENRT/UseVfsMixin.py (hunk 1 above)
Comment (jtluka) [reply lnst/Recipes/ENRT/UseVfsMixin.py:33]: I changed this to be able to access the vf_config in `test_wide_description()`. I can add this to both `config.vf_config()` and `self.vf_config()`

Comment (jtluka) [reply lnst/Recipes/ENRT/UseVfsMixin.py:33]: Err, and when I checked the code now I see the config is passed as argument. I'll double check that, most likely you have a valid point.

File: lnst/Recipes/ENRT/UseVfsMixin.py (hunk 2 above)
Comment (olichtne) [reply lnst/Recipes/ENRT/UseVfsMixin.py:49]: ah ok, it doesn't have a default, and parameters are optional by default... ok

File: lnst/Recipes/ENRT/UseVfsMixin.py (hunk 1 above)
Comment (jtluka) [reply lnst/Recipes/ENRT/UseVfsMixin.py:33]: I changed this because I need this in https://github.com/LNST-project/lnst/blob/1d72d11f371312ac835c373b1d539e4757bf32eb/lnst/Recipes/ENRT/BondRecipe.py#L252.

To have the device list as property I need to store this in recipe object directly.

File: lnst/Recipes/ENRT/BondRecipe.py (hunk 4 above)
Comment (jtluka) [reply lnst/Recipes/ENRT/BondRecipe.py:252]: No, this is not necessary. SriovDevices is a collection of VFs over **single** PF. So iterating over devices in sriov_devices would generate several PFs that are identical.

Comment (jtluka) [reply lnst/Recipes/ENRT/BondRecipe.py:252]: And `SriovDevices` class exports the `phys_dev` property that is common to all VFs in the class.

File: lnst/Recipes/ENRT/DoubleBondRecipe.py (hunk 5 above)
Comment (jtluka) [reply lnst/Recipes/ENRT/DoubleBondRecipe.py:125]: The class name is `SRIOVDevices`, so I followed this pattern. I thought of SRIOVDevices class as a collection of VFs over a PF.

The `self.vf_config.values()` returns `list[list[SRIOVDevices]]`.

Comment (olichtne) [reply lnst/Recipes/ENRT/DoubleBondRecipe.py:125]: > The self.vf_config.values() returns list[list[SRIOVDevices]].

this is why i'm confused about this `vf_trust_device_list` property, in some recipes you have two nested loops and in some just one loop...

File: lnst/Recipes/ENRT/UseVfsMixin.py (hunk 1 above)
Comment (jtluka) [reply lnst/Recipes/ENRT/UseVfsMixin.py]: Latest version does not require this anymore, so I removed this change.

File: lnst/Recipes/ENRT/DoubleBondRecipe.py (hunk 5 above)
Comment (jtluka) [reply lnst/Recipes/ENRT/DoubleBondRecipe.py]: I changed this so that the individual recipes return list of the vf_devices (the remapped device requirements) and the UseVfsMixin that has access to SRIOVDevices structures would gather the PF devices.

File: lnst/Recipes/ENRT/VlansOverBondRecipe.py (hunk 6 above)
Comment (jtluka) [reply lnst/Recipes/ENRT/VlansOverBondRecipe.py]: Updated per https://github.com/LNST-project/lnst/pull/381#discussion_r1805965954

File: lnst/Recipes/ENRT/BondRecipe.py (hunk 4 above)
Comment (jtluka) [reply lnst/Recipes/ENRT/BondRecipe.py]: Updated per https://github.com/LNST-project/lnst/pull/381#discussion_r1805965954

**Review by olichtne — APPROVED** [review]
just a minor nitpick, otherwise approved

File: lnst/Recipes/ENRT/BondingMixin.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        )
+
+        if self.params.bonding_mode in ["active-backup", "1"]:
+            device_params["fail_over_mac"] = self.params.fail_over_mac
+
+        for host_str, spec in bond_devices_specs.items():
+            for bond_dev_name, bonded_devices in spec.items():
+                if len({dev.host for dev in bonded_devices}) > 1:
+                    raise RecipeError(
+                        f"Cannot create bond device with ports coming from different hosts, {bonded_devices}"
```
[the remark this answers is not in the document]
Comment (olichtne) [reply lnst/Recipes/ENRT/BondingMixin.py:62]: the condition you use counts the number of hosts in "bonded_devices", if there's 2 then yes... the devices are from two different hosts and that doesn't work...

however, you can also provide 2 devices from one host, which is different than the host where the bond device should be created - your condition will count 1, but will pass even when it should fail.

for example:
```
        self.create_bond_devices(
            {
                "host1": {
                    "bond0": [host2.nic1, host2.nic2]
                },
            }
```

nics are from host2, but bond on host1.

[the remark this answers is not in the document]
Comment (jtluka) [reply lnst/Recipes/ENRT/BondingMixin.py:62]: I'm not referring to the host_str in the spec. I'm just checking if all devices in the bond_devices list are on the same host.

[the remark this answers is not in the document]
Comment (jtluka) [reply lnst/Recipes/ENRT/BondingMixin.py:62]: I could add your condition with a different exception string if you want.

[the remark this answers is not in the document]
Comment (olichtne) [reply lnst/Recipes/ENRT/BondingMixin.py:62]: > if all devices in the bond_devices list are on the same host.

yes... but it could be the WRONG host. I know you're not checking `host_str`, i'm saying maybe we should?

but yeah... it's a nitpick like i've said, we can skip this if you want.

[the remark this answers is not in the document]
Comment (jtluka) [reply lnst/Recipes/ENRT/BondingMixin.py:62]: You're right that with your proposal that would cover both cases. I'll do some testing and eventually update this.

[the remark this answers is not in the document]
Comment (jtluka) [reply lnst/Recipes/ENRT/BondingMixin.py]: Fixed, I've also added check if the `host_str` can be resolved in `self.matched.hosts`

**Review by olichtne — DISMISSED** [review]

File: lnst/Recipes/ENRT/BondingMixin.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (olichtne) [reply lnst/Recipes/ENRT/BondingMixin.py]: looks ok

**Comment by olichtne** [comment]
add doc rst file entries so that the new classes are exposed on readthedocs properly

File: lnst/Recipes/ENRT/VlansOverBondRecipe.py [hunk 8]
```diff
@@ -200,6 +192,12 @@ def generate_test_wide_description(self, config: EnrtConfiguration):
                 host1.bond0.miimon
             )
         ]
+
+        if self.params.bonding_mode in ["active-backup", "1"]:
+            desc += ["Configured {}.{}.fail_over_mac = {}".format(
+                host1.hostid, host1.bond0.name,
+                host1.bond0.fail_over_mac
+            )]
```
Comment (jtluka) [inline lnst/Recipes/ENRT/VlansOverBondRecipe.py:200]: I think this shoud go into  the BondingMixin class. It's duplicated over individual recipes and it makes more sense to generate this within the BondingMixin class.

Comment (jtluka) [reply lnst/Recipes/ENRT/VlansOverBondRecipe.py]: Fixed.

**Comment by jtluka** [comment]
Last test for fail_over_mac setting : J:10057391

**Comment by jtluka** [comment]
> Last test for fail_over_mac setting : J:10057391

Tests passed.
